// node-hl7-client/lib/cjs/builder/modules/emptyNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmptyNode = void 0;
/**
 * Empty Node
 * @since 1.0.0
 */
class EmptyNode {
    get name() {
        throw new Error("Method not implemented");
    }
    get length() {
        return 0;
    }
    get(_path) {
        return this;
    }
    set(_path, _value) {
        return this;
    }
    exists(_path) {
        return false;
    }
    forEach(_callback) {
        throw new Error("Method not implemented");
    }
    toString() {
        return "";
    }
    toFile(_name, _newLine, _location) {
        throw new Error("Method not implemented.");
    }
    toRaw() {
        throw new Error("Method not implemented");
    }
    toArray() {
        return [];
    }
    isEmpty() {
        return true;
    }
    toDate() {
        throw new Error("Method not implemented");
    }
    toInteger() {
        throw new Error("Method not implemented");
    }
    toFloat() {
        throw new Error("Method not implemented");
    }
    toBoolean() {
        throw new Error("Method not implemented");
    }
    read(_path) {
        throw new Error("Method not implemented");
    }
    write(_path, _value) {
        return this;
    }
    get path() {
        throw new Error("Method not implemented");
    }
}
exports.EmptyNode = EmptyNode;
