// node-hl7-client/lib/cjs/builder/modules/nodeBase.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeBase = void 0;
const events_1 = __importDefault(require("events"));
const utils_js_1 = require("../../utils/utils.js");
const emptyNode_js_1 = require("./emptyNode.js");
const exception_js_1 = require("../../utils/exception.js");
/**
 * Node Base
 * @since 1.0.0
 */
class NodeBase extends events_1.default {
    constructor(parent, text = "", delimiter = undefined) {
        super();
        this.parent = parent;
        this._children = [];
        this._delimiter = delimiter;
        this._delimiterText = "";
        this._dirty = false;
        this._name = "";
        this._path = [];
        this._text = text;
    }
    get(path) {
        let ret;
        if (typeof path === "number") {
            if (path >= 0 && path < this.children.length) {
                ret = this.children[path];
            }
        }
        else if (path !== "") {
            const _path = this.preparePath(path);
            ret = this.read(_path);
        }
        return typeof ret !== "undefined"
            ? ret
            : NodeBase.empty;
    }
    set(path, value) {
        if (arguments.length === 1) {
            return this.ensure(path);
        }
        if (typeof path === "string") {
            if (Array.isArray(value)) {
                for (let i = 0; i < value.length; i++) {
                    this.set(`${path}.${i + 1}`, value[i]);
                }
            }
            else {
                const _path = this.preparePath(path);
                this.write(_path, this.prepareValue(value));
            }
            return this;
        }
        else if ((0, utils_js_1.isHL7Number)(path)) {
            if (Array.isArray(value)) {
                const child = this.ensure(path);
                for (let i = 0, l = value.length; i < l; i++) {
                    child.set(i, value[i]);
                }
                return this;
            }
            else {
                this.setChild(this.createChild(this.prepareValue(value), path), path);
            }
            return this;
        }
        throw new exception_js_1.HL7FatalError("Path must be a string or number.");
    }
    get name() {
        if (this._name !== undefined) {
            return this._name;
        }
        this._name = this.path.join(".");
        return this._name;
    }
    get length() {
        return this.children.length;
    }
    toDate() {
        throw new Error("Method not implemented.");
    }
    toFile(_name, _newLine, _location) {
        throw new Error("Method not implemented.");
    }
    toInteger() {
        throw new Error("Method not implemented.");
    }
    toFloat() {
        throw new Error("Method not implemented.");
    }
    toBoolean() {
        throw new Error("Method not implemented.");
    }
    toString() {
        return this.toRaw();
    }
    toRaw() {
        if (!this._dirty) {
            return typeof this._text !== "undefined" ? this._text : "";
        }
        this._dirty = false;
        this._text = this.children.map((x) => x.toRaw()).join(this.delimiter);
        return this._text;
    }
    toArray() {
        return this.children;
    }
    forEach(callback) {
        const children = this.children;
        for (let i = 0, l = children.length; i < l; i++) {
            callback(children[i], i);
        }
    }
    exists(path) {
        const value = this.get(path);
        if (value == null) {
            return false;
        }
        return !value.isEmpty();
    }
    isEmpty() {
        return this.children.length === 0;
    }
    /** @internal */
    ensure(path) {
        const ret = this.get(path);
        if (ret !== NodeBase.empty) {
            return ret;
        }
        if (typeof path === "number") {
            return this.setChild(this.createChild("", path), path);
        }
        else if ((0, utils_js_1.isHL7String)(path)) {
            return this.write(this.preparePath(path), "");
        }
        throw new exception_js_1.HL7FatalError("There seems to be a problem.");
    }
    /** @internal */
    preparePath(path) {
        let parts = path.split(".");
        if (parts[0] === "") {
            parts.shift();
            parts = this.path.concat(parts);
        }
        if (!this._isSubPath(parts)) {
            throw new exception_js_1.HL7FatalError("'" +
                parts.toString() +
                "' is not a sub-path of '" +
                this.path.toString() +
                "'");
        }
        return this._remainderOf(parts);
    }
    /** @internal */
    prepareValue(value) {
        if (value == null)
            return "";
        if (typeof value === "string") {
            if (typeof this.message !== "undefined") {
                return this.message.escape(value);
            }
        }
        if (typeof value === "number") {
            return value.toString();
        }
        if (typeof value === "boolean") {
            return value ? "Y" : "N";
        }
        if (value instanceof Date) {
            return this._formatDateTime(value);
        }
        return value.toString();
    }
    /** @internal */
    get message() {
        if (typeof this._message !== "undefined") {
            return this._message;
        }
        this._message = this.parent !== null ? this.parent.message : this;
        return this._message;
    }
    read(_path) {
        throw new Error("Method not implemented.");
    }
    write(path, value) {
        this.setDirty();
        return this.writeCore(path, value == null ? "" : value);
    }
    /** @internal */
    writeCore(_path, _value) {
        throw new Error("Method not implemented.");
    }
    /** @internal */
    writeAtIndex(path, value, index, emptyValue = "") {
        let child;
        if (path.length === 0) {
            child = this.createChild(typeof value !== "undefined" ? value : emptyValue, index);
        }
        else {
            if (index < this.children.length) {
                child = this.children[index];
            }
            else {
                child = this.createChild(emptyValue, index);
            }
        }
        this.setChild(child, index);
        if (path.length !== 0) {
            return child.write(path, value);
        }
        return child;
    }
    get path() {
        if (typeof this._path !== "undefined") {
            return this._path;
        }
        this._path = this.pathCore();
        return this._path;
    }
    /** @internal */
    pathCore() {
        throw new Error("Method not implemented.");
    }
    /** @internal */
    get delimiter() {
        if (typeof this.message === "undefined") {
            throw new exception_js_1.HL7FatalError("this.message is not defined.");
        }
        if (typeof this._delimiter === "undefined") {
            throw new exception_js_1.HL7FatalError("this.message is not defined.");
        }
        this._delimiterText = this.message.delimiters[this._delimiter];
        return this._delimiterText;
    }
    /** @internal */
    get children() {
        if (this._text !== "" && this._children.length === 0) {
            const parts = this._text.split(this.delimiter);
            const children = new Array(parts.length);
            for (let i = 0, l = parts.length; i < l; i++) {
                children[i] = this.createChild(parts[i], i);
            }
            this._children = children;
        }
        return this._children;
    }
    /** @internal */
    addChild(text) {
        this.setDirty();
        const child = this.createChild(text, this.children.length);
        this.children.push(child);
        return child;
    }
    /** @internal */
    createChild(_text, _index) {
        throw new Error("Method not implemented.");
    }
    /** @internal */
    setChild(child, index) {
        this.setDirty();
        const children = this.children;
        if (index < children.length) {
            children[index] = child;
            return child;
        }
        for (let i = children.length; i < index; i++) {
            children.push(this.createChild("", i));
        }
        children.push(child);
        return child;
    }
    /** @internal */
    setDirty() {
        if (!this._dirty) {
            this._dirty = true;
            if (typeof this.parent !== "undefined" && this.parent !== null) {
                this.parent.setDirty();
            }
        }
    }
    /** @internal */
    _isSubPath(other) {
        if (this.path.length >= other.length) {
            return false;
        }
        const path = this.path;
        for (let i = 0, l = path.length; i < l; i++) {
            if (path[i] !== other[i]) {
                return false;
            }
        }
        return true;
    }
    /** @internal */
    _remainderOf(other) {
        const path = this.path;
        return other.slice(path.length);
    }
    /** @internal */
    _formatDateTime(date) {
        // check if there is a time component
        if (date.getHours() !== 0 ||
            date.getMinutes() !== 0 ||
            date.getSeconds() !== 0 ||
            date.getMilliseconds() !== 0) {
            return (this._formatDate(date) +
                (0, utils_js_1.padHL7Date)(date.getHours(), 2) +
                (0, utils_js_1.padHL7Date)(date.getMinutes(), 2) +
                (0, utils_js_1.padHL7Date)(date.getSeconds(), 2));
        }
        return this._formatDate(date);
    }
    /** @internal */
    _formatDate(date) {
        return (date.getFullYear().toString() +
            (0, utils_js_1.padHL7Date)(date.getMonth() + 1, 2) +
            (0, utils_js_1.padHL7Date)(date.getDate(), 2));
    }
}
exports.NodeBase = NodeBase;
NodeBase.empty = new emptyNode_js_1.EmptyNode();
