// node-hl7-client/lib/cjs/builder/modules/valueNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValueNode = void 0;
const exception_js_1 = require("../../utils/exception.js");
const nodeBase_js_1 = require("./nodeBase.js");
/** @internal */
class ValueNode extends nodeBase_js_1.NodeBase {
    /** @internal */
    constructor(parent, key, text, delimiter = undefined) {
        super(parent, text, delimiter);
        this.key = key;
    }
    /** @internal */
    toString() {
        const children = this.children;
        return children.length === 0 ? this.toRaw() : children[0].toString();
    }
    /** @internal */
    toDate() {
        const text = this.toString();
        if (text.length === 8) {
            /** YYYYMMDD */
            return new Date(parseInt(text.slice(0, 4)), parseInt(text.slice(4, 6)) - 1, parseInt(text.slice(6, 8)), 0, 0, 0);
        }
        else if (text.length === 12) {
            /** YYYYMMDDHHMM */
            return new Date(parseInt(text.slice(0, 4)), parseInt(text.slice(4, 6)) - 1, parseInt(text.slice(6, 8)), parseInt(text.slice(8, 10)), parseInt(text.slice(10, 12)), 0);
        }
        else if (text.length >= 14) {
            /** YYYYMMDDHHMMSS **/
            return new Date(parseInt(text.slice(0, 4)), parseInt(text.slice(4, 6)) - 1, parseInt(text.slice(6, 8)), parseInt(text.slice(8, 10)), parseInt(text.slice(10, 12)), parseInt(text.slice(12, 14)));
        }
        throw new exception_js_1.HL7FatalError("Invalid Date Format");
    }
    /** @internal */
    toInteger() {
        return parseInt(this.toString());
    }
    /** @internal */
    toFloat() {
        return parseFloat(this.toString());
    }
    /** @internal */
    toBoolean() {
        switch (this.toString()) {
            case "Y":
                return true;
            case "N":
                return false;
        }
        throw new exception_js_1.HL7FatalError("Not a valid value for boolean value.");
    }
    /** @internal */
    pathCore() {
        if (this.parent === null) {
            throw new exception_js_1.HL7FatalError("Somehow, this.parent is null.");
        }
        return this.parent.path.concat([this.key]);
    }
}
exports.ValueNode = ValueNode;
