// node-hl7-client/lib/cjs/client/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const events_1 = __importDefault(require("events"));
const normalizedClient_js_1 = require("../utils/normalizedClient.js");
const connection_js_1 = require("./connection.js");
/**
 * Client Class
 * @remarks The main class that starts a client connection to a valid HL7 TCP/MLLP specified server.
 * @since 1.0.0 */
class Client extends events_1.default {
    /**
     * This creates a new client to a new server connection.
     * This is the remote side in which we will connect.
     * Then using the {@link createConnection} method, you establish a connection to the port.
     * @since 1.0.0
     * @param props
     * @example
     * ```ts
     * const client = new Client({host: '0.0.0.0'})
     * ```
     */
    constructor(props) {
        super();
        /** @internal */
        this.stats = {
            /** Total outbound connections able to connect to at this moment.
             * @since 1.1.0 */
            _totalConnections: 0,
            /** Overall total sent messages
             * @since 2.0.0 */
            _totalSent: 0,
            /** Overall Ack *
             * @since 2.0.0 */
            _totalAck: 0,
        };
        this._opt = (0, normalizedClient_js_1.normalizeClientOptions)(props);
        this._connections = [];
    }
    /**
     * Close all connections
     * @since 2.0.0
     */
    closeAll() {
        // loop through!
        this._connections.map(async (connection) => {
            void connection.close();
        });
        // reset!
        this._connections = [];
    }
    /** Connect to a listener to a specified port.
     * @since 1.0.0
     * @param props This individual port connections which can override the main server connection properties.
     * Some properties from the server build could be defaulted if not specified here.
     * @param cb The function that the client will process if and when they get a response from the server.
     * It follows an async/await function.
     * @example
     * ```ts
     * const outGoing = client.createOutbound({port: 3000}, async (res: InboundResponse) => {})
     * ```
     * Review the {@link InboundResponse} on the properties returned.
     */
    createConnection(props, cb) {
        const outbound = new connection_js_1.Connection(this, props, cb);
        outbound.on("client.acknowledged", (total) => {
            this.stats._totalAck = total;
        });
        outbound.on("client.sent", (total) => {
            this.stats._totalSent = total;
        });
        // add this connection
        this._connections.push(outbound);
        // send back current outbound
        return outbound;
    }
    /**
     * Get the host that we will connect to.
     * The port might be different from each different "connection"
     * @since 1.1.0
     */
    getHost() {
        return this._opt.host;
    }
    /**
     * Total ack in this object lifetime.
     * @since 2.0.0
     */
    totalAck() {
        return this.stats._totalAck;
    }
    /**
     * Total sent messages in this object lifetime.
     * @since 2.0.0
     */
    totalSent() {
        return this.stats._totalSent;
    }
}
exports.Client = Client;
