// node-hl7-client/lib/cjs/client/connection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connection = void 0;
const node_events_1 = __importDefault(require("node:events"));
const node_net_1 = __importDefault(require("node:net"));
const node_timers_1 = require("node:timers");
const node_tls_1 = __importDefault(require("node:tls"));
const batch_js_1 = __importDefault(require("../builder/batch.js"));
const message_js_1 = __importDefault(require("../builder/message.js"));
const enum_js_1 = require("../utils/enum.js");
const exception_js_1 = require("../utils/exception.js");
const normalizedClient_js_1 = require("../utils/normalizedClient.js");
const utils_js_1 = require("../utils/utils.js");
const inboundResponse_js_1 = require("./module/inboundResponse.js");
const codec_js_1 = require("../utils/codec.js");
/** Connection Class
 * @remarks Create a connection customer that will listen to result send to the particular port.
 * @since 1.0.0 */
class Connection extends node_events_1.default {
    /**
     * @since 1.0.0
     * @param client The client parent that we are connecting too.
     * @param props The individual port connection options.
     * Some values will be defaulted by the parent server connection.
     * @param handler The function that will send the returned information back to the client after we got a response from the server.
     * @example
     * ```ts
     * const OB = client.createConnection({ port: 3000 }, async (res) => {})
     * ```
     */
    constructor(client, props, handler) {
        super();
        /** @internal */
        this.stats = {
            /** Total acknowledged messages back from server.
             * @since 1.1.0 */
            acknowledged: 0,
            /** Total message sent to server.
             * @since 1.1.0 */
            sent: 0,
        };
        this._handler = handler;
        this._main = client;
        this._awaitingResponse = false;
        this._opt = (0, normalizedClient_js_1.normalizeClientListenerOptions)(client._opt, props);
        this._connect = this._connect.bind(this);
        this._pendingSetup = true;
        this._retryCount = 0;
        this._retryTimeoutCount = 0;
        this._retryTimer = undefined;
        this._connectionTimer = undefined;
        this._codec = null;
        this._onConnect = (0, utils_js_1.createDeferred)(true);
        if (this._opt.autoConnect) {
            this._readyState = enum_js_1.ReadyState.CONNECTING;
            this.emit("connecting");
            this._socket = this._connect();
        }
        else {
            this._readyState = enum_js_1.ReadyState.OPEN;
            this.emit("open");
            this._socket = undefined;
        }
    }
    /** Close Client Listener Instance.
     * @remarks Force close a connection.
     * It will stop any re-connection timers.
     * If you want to restart, your app has to restart the connection.
     * @since 1.0.0
     * @example
     * ```ts
     * OB.close()
     * ```
     */
    async close() {
        if (this._readyState === enum_js_1.ReadyState.CLOSED) {
            return; // We are already closed. Nothing to do.
        }
        if (this._readyState === enum_js_1.ReadyState.CLOSING) {
            return await new Promise((resolve) => this._socket?.once("close", resolve));
        }
        if (this._readyState === enum_js_1.ReadyState.CONNECTING) {
            // clear retry timer
            if (typeof this._retryTimer !== "undefined") {
                (0, node_timers_1.clearTimeout)(this._retryTimer);
            }
            // let's clear out the try timer forcefully
            this._retryTimer = undefined;
        }
        // normal closing
        this._readyState = enum_js_1.ReadyState.CLOSING;
        // remove socket
        this._socket?.destroy();
        this._socket?.end();
        // Ensure no further reconnections are attempted
        (0, node_timers_1.clearTimeout)(this._connectionTimer);
        // Emit close event on the socket
        this.emit("close");
        // Set closure state
        this._readyState = enum_js_1.ReadyState.CLOSED;
    }
    /**
     * Get Port
     * @remarks Get the port that this connection will connect to.
     * @since 2.0.0
     */
    getPort() {
        return this._opt.port;
    }
    /**
     * Start the connection if not auto started.
     * @since 2.0.0
     */
    async start() {
        if (this._readyState === enum_js_1.ReadyState.CONNECTING) {
            return;
        }
        if (this._readyState === enum_js_1.ReadyState.CONNECTED) {
            return;
        }
        if (this._readyState === enum_js_1.ReadyState.OPEN) {
            return;
        }
        if (this._readyState === enum_js_1.ReadyState.CLOSING) {
            return await new Promise((resolve) => this._socket?.once("close", resolve));
        }
        this.emit("connecting");
        this._socket = this._connect();
    }
    /** Send a HL7 Message to the Listener
     * @remarks This function sends a message/batch/file batch to the remote side.
     * It has the ability, if set to auto-retry (defaulted to 1 re-connect before connection closes)
     * @since 1.0.0
     * @param message The message we need to send to the port.
     * @example
     * ```ts
     *
     * // the OB was set from the orginial 'createConnection' method.
     *
     * let message = new Message({
     *  messageHeader: {
     *    msh_9_1: "ADT",
     *    msh_9_2: "A01",
     *    msh_11_1: "P" // marked for production here in the example
     *  }async sendMessage (message: Message | Batch | FileBatch): void {
     * })
     *
     * await OB.sendMessage(message)
     *
     * ```
     */
    async sendMessage(message) {
        let attempts = 0;
        const maxAttempts = this._opt.maxAttempts;
        const emitter = new node_events_1.default();
        const codec = new codec_js_1.MLLPCodec(this._opt.encoding);
        const checkConnection = async () => {
            return this._readyState === enum_js_1.ReadyState.CONNECTED;
        };
        const checkAcknowledgement = async () => {
            return this._awaitingResponse;
        };
        const checkSend = async (_message) => {
            while (true) {
                // noinspection InfiniteLoopJS
                try {
                    if (this._readyState === enum_js_1.ReadyState.CLOSED ||
                        this._readyState === enum_js_1.ReadyState.CLOSING) {
                        // noinspection ExceptionCaughtLocallyJS
                        throw new exception_js_1.HL7FatalError("In an invalid state to be able to send message.");
                    }
                    if (this._readyState !== enum_js_1.ReadyState.CONNECTED) {
                        // if we are not connected,
                        // check to see if we are now connected.
                        if (this._pendingSetup === false) {
                            this._pendingSetup = checkConnection().finally(() => {
                                this._pendingSetup = false;
                            });
                        }
                    }
                    else if (this._readyState === enum_js_1.ReadyState.CONNECTED &&
                        this._opt.waitAck &&
                        this._awaitingResponse) {
                        // Ok, we ar now confirmed connected.
                        // However, since we are checking
                        // to make sure we wait for an ACKNOWLEDGEMENT from the server,
                        // that the message was gotten correctly from the last one we sent.
                        // We are still waiting, we need to recheck again
                        // if we are not connected,
                        // check to see if we are now connected.
                        if (this._pendingSetup === false) {
                            this._pendingSetup = checkAcknowledgement().finally(() => {
                                this._pendingSetup = false;
                            });
                        }
                    }
                    return await this._pendingSetup;
                }
                catch (err) {
                    Error.captureStackTrace(err);
                    if (++attempts >= maxAttempts) {
                        throw err;
                    }
                    else {
                        emitter.emit("retry", err);
                    }
                }
            }
        };
        // get the message
        const theMessage = message.toString();
        // check to see if we should be sending
        await checkSend(theMessage);
        // ok, if our options are to wait for an acknowledgement, set the var to "true"
        if (this._opt.waitAck) {
            this._awaitingResponse = true;
        }
        // send the message
        codec.sendMessage(this._socket, theMessage, this._opt.encoding);
        // we sent a message
        ++this.stats.sent;
        // emit
        this.emit("client.sent", this.stats.sent);
    }
    /** @internal */
    _connect() {
        let socket;
        const host = this._main._opt.host;
        const port = this._opt.port;
        this._retryTimer = undefined;
        if (typeof this._main._opt.tls !== "undefined") {
            socket = node_tls_1.default.connect({
                host,
                port,
                ...this._main._opt.socket,
                ...this._main._opt.tls,
            });
        }
        else {
            socket = node_net_1.default.connect({
                host,
                port,
            });
        }
        this._codec = new codec_js_1.MLLPCodec();
        this._socket = socket;
        // set no delay
        socket.setNoDelay(true);
        let connectionError;
        if (this._main._opt.connectionTimeout > 0 &&
            (this._readyState === enum_js_1.ReadyState.CONNECTED ||
                this._readyState === enum_js_1.ReadyState.CONNECTING)) {
            if (this._retryTimeoutCount < this._main._opt.maxTimeout) {
                this._connectionTimer = setTimeout(() => {
                    ++this._retryTimeoutCount;
                    this.emit("client.timeout");
                    socket.destroy();
                }, this._main._opt.connectionTimeout);
            }
            else if (this._retryTimeoutCount >= this._main._opt.maxTimeout) {
                void this.close();
            }
        }
        socket.on("error", (err) => {
            connectionError = connectionError != null ? connectionError : err;
        });
        socket.on("close", () => {
            if (this._readyState === enum_js_1.ReadyState.CLOSING) {
                this._readyState = enum_js_1.ReadyState.CLOSED;
            }
            else if (!(this._readyState === enum_js_1.ReadyState.CLOSED &&
                this._connectionTimer != null &&
                this._connectionTimer
                    ._destroyed) &&
                this._main._opt.connectionTimeout > 0) {
                connectionError =
                    connectionError != null
                        ? connectionError
                        : new exception_js_1.HL7FatalError("Socket closed unexpectedly by server.");
                if (this._readyState === enum_js_1.ReadyState.OPEN) {
                    this._onConnect = (0, utils_js_1.createDeferred)(true);
                }
                this._readyState = enum_js_1.ReadyState.CONNECTING;
                const retryCount = this._retryCount++;
                const delay = (0, utils_js_1.expBackoff)(this._opt.retryLow, this._opt.retryHigh, retryCount);
                if (retryCount < this._opt.maxConnectionAttempts) {
                    this._retryTimer = setTimeout(this._connect, delay);
                    this.emit("client.error", connectionError);
                }
                else if (retryCount > this._opt.maxConnectionAttempts) {
                    // stop this from going again
                    void this.close();
                }
            }
        });
        socket.on("connect", () => {
            // accepting connections
            this._readyState = enum_js_1.ReadyState.CONNECTED;
            // reset retryCount count
            this._retryCount = 1;
            // emit
            this.emit("connect");
        });
        socket.on("data", (buffer) => {
            socket.cork();
            try {
                this._dataResult = this._codec?.receiveData(buffer);
            }
            catch (err) {
                this.emit("data.error", err);
            }
            socket.uncork();
            if (this._dataResult === true) {
                // we got some sort of response, bad, good, or error,
                // so let's tell the system we got "something"
                this._awaitingResponse = false;
                try {
                    const loadedMessage = this._codec?.getLastMessage();
                    // copy the completed message to continue processing and clear the buffer
                    const completedMessageCopy = JSON.parse(JSON.stringify(loadedMessage));
                    // parser either is batch or a message
                    let parser;
                    // send raw information to the emit
                    this.emit("data.raw", completedMessageCopy);
                    if (typeof this._handler !== "undefined") {
                        if ((0, utils_js_1.isBatch)(completedMessageCopy)) {
                            // parser the batch
                            parser = new batch_js_1.default({ text: completedMessageCopy });
                            // load the messages
                            const allMessage = parser.messages();
                            // loop messages
                            allMessage.forEach((message) => {
                                // parse this message
                                const messageParsed = new message_js_1.default({ text: message.toString() });
                                // increase the total message
                                ++this.stats.acknowledged;
                                // response
                                const response = new inboundResponse_js_1.InboundResponse(messageParsed.toString());
                                // update ack total
                                this.emit("client.acknowledged", this.stats.acknowledged);
                                // send it back
                                void this._handler(response);
                            });
                        }
                        else {
                            // parse this message
                            const messageParsed = new message_js_1.default({ text: completedMessageCopy });
                            // increase the total message
                            ++this.stats.acknowledged;
                            // response
                            const response = new inboundResponse_js_1.InboundResponse(messageParsed.toString());
                            // update ack total
                            this.emit("client.acknowledged", this.stats.acknowledged);
                            // send it back
                            void this._handler(response);
                        }
                    }
                }
                catch (err) {
                    this.emit("data.error", err);
                }
            }
        });
        const readerLoop = async () => {
            try {
                await this._negotiate();
            }
            catch (err) {
                if (err.code !== "READ_END") {
                    socket.destroy(err);
                }
            }
        };
        void readerLoop().then((_r) => {
            /* noop */
        });
        return socket;
    }
    /** @internal */
    async _negotiate() {
        if (this._socket?.writable === true) {
            // we are open, not yet ready, but we can
            this._readyState = enum_js_1.ReadyState.OPEN;
            // on connect resolve
            this._onConnect.resolve();
            // emit
            this.emit("connection");
        }
    }
}
exports.Connection = Connection;
exports.default = Connection;
