// node-hl7-client/lib/esm/utils/enum.js
/**
 * Type of Segments Values
 * @remarks Used during the class creation to give each type its own index value.
 * This is done during the constructor phase of the classes.
 * @since 1.0.0
 */
export var Delimiters;
(function (Delimiters) {
    /** Usually each line of the overall HL7 Message. */
    Delimiters[Delimiters["Segment"] = 0] = "Segment";
    /** The field of each segment. Usually separated with a | */
    Delimiters[Delimiters["Field"] = 1] = "Field";
    /** Usually within each Field, seperated by ^ */
    Delimiters[Delimiters["Component"] = 2] = "Component";
    /** Usually within each Component, seperated by & */
    Delimiters[Delimiters["Repetition"] = 3] = "Repetition";
    /** The escape string used within the code. */
    Delimiters[Delimiters["Escape"] = 4] = "Escape";
    /** Usually within each Field, seperated by ~ */
    Delimiters[Delimiters["SubComponent"] = 5] = "SubComponent";
})(Delimiters || (Delimiters = {}));
/**
 * State of the Connected to the Server
 * @remarks These are the states that are used to track the connecting to the server side and also during the auto-reconnect phase.
 * @since 1.0.0
 */
export var ReadyState;
(function (ReadyState) {
    /** The client is trying to connect to the server. */
    ReadyState[ReadyState["CONNECTING"] = 0] = "CONNECTING";
    /** The client is connected to the server.  */
    ReadyState[ReadyState["CONNECTED"] = 1] = "CONNECTED";
    /** The client is open, but not yet trying to connect to the server.  */
    ReadyState[ReadyState["OPEN"] = 2] = "OPEN";
    /** The client is closing the connection by force or by timeout */
    ReadyState[ReadyState["CLOSING"] = 3] = "CLOSING";
    /** The client connection is closed.  */
    ReadyState[ReadyState["CLOSED"] = 4] = "CLOSED";
})(ReadyState || (ReadyState = {}));
