// node-hl7-client/lib/esm/utils/normalizedClient.js
import { HL7FatalError } from "./exception.js";
import { assertNumber, validIPv4, validIPv6 } from "./utils.js";
const DEFAULT_CLIENT_OPTS = {
    encoding: "utf-8",
    connectionTimeout: 0,
    maxAttempts: 10,
    maxConnectionAttempts: 10,
    maxTimeout: 10,
    retryHigh: 30000,
    retryLow: 1000,
};
const DEFAULT_LISTEN_CLIENT_OPTS = {
    autoConnect: true,
    maxAttempts: 10,
    maxConnectionAttempts: 10,
    waitAck: true,
};
/** @internal */
export function normalizeClientOptions(raw) {
    const props = { ...DEFAULT_CLIENT_OPTS, ...raw };
    if (typeof props.host === "undefined" || props.host.length <= 0) {
        throw new HL7FatalError("host is not defined or the length is less than 0.");
    }
    if (props.ipv4 === true && props.ipv6 === true) {
        throw new HL7FatalError("ipv4 and ipv6 both can't be set to be both used exclusively.");
    }
    if (typeof props.host !== "string" &&
        props.ipv4 === false &&
        props.ipv6 === false) {
        throw new HL7FatalError("host is not valid string.");
    }
    else if (typeof props.host === "string" &&
        props.ipv4 === true &&
        props.ipv6 === false) {
        if (!validIPv4(props.host)) {
            throw new HL7FatalError("host is not a valid IPv4 address.");
        }
    }
    else if (typeof props.host === "string" &&
        props.ipv4 === false &&
        props.ipv6 === true) {
        if (!validIPv6(props.host)) {
            throw new HL7FatalError("host is not a valid IPv6 address.");
        }
    }
    if (props.tls === true) {
        props.tls = {};
    }
    assertNumber(props, "connectionTimeout", 0, 60000);
    assertNumber(props, "maxTimeout", 1, 50);
    return props;
}
/** @internal */
export function normalizeClientListenerOptions(client, raw) {
    const props = { ...DEFAULT_LISTEN_CLIENT_OPTS, ...raw };
    if (typeof props.port === "undefined") {
        throw new HL7FatalError("port is not defined.");
    }
    if (typeof props.port !== "number") {
        throw new HL7FatalError("port is not valid number.");
    }
    if (typeof props.retryHigh === "undefined") {
        props.retryHigh = client.retryHigh;
    }
    if (typeof props.retryLow === "undefined") {
        props.retryLow = client.retryLow;
    }
    assertNumber(props, "maxAttempts", 1, 50);
    assertNumber(props, "maxConnectionAttempts", 1, 50);
    assertNumber(props, "port", 1, 65353);
    return props;
}
