// node-hl7-client/lib/cjs/builder/fileBatch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileBatch = void 0;
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
const constants_js_1 = require("../utils/constants.js");
const exception_js_1 = require("../utils/exception.js");
const normalizedBuilder_js_1 = require("../utils/normalizedBuilder.js");
const utils_js_1 = require("../utils/utils.js");
const batch_js_1 = require("./batch.js");
const message_js_1 = require("./message.js");
const rootBase_js_1 = require("./modules/rootBase.js");
const segment_js_1 = require("./modules/segment.js");
const segmentList_js_1 = require("./modules/segmentList.js");
/**
 * File Batch Class
 * @remarks Create a File Batch (FHS) which will could include many BHS/BTS segments,
 * which could include many Message (MSH) segments to output the contents into a file on the OS.
 * These files could then be used to send manually or read by another system to interpret the contents.
 * This class helps
 * in generating the particulars for that file generation to make sure that it follows the correct format.
 * @since 1.0.0
 */
class FileBatch extends rootBase_js_1.RootBase {
    /**
     * @since 1.0.0
     * @param props Passing the options to build the file batch.
     */
    constructor(props) {
        const opt = (0, normalizedBuilder_js_1.normalizedClientFileBuilderOptions)(props);
        super(opt);
        this._fileName = "";
        this._lines = [];
        this._opt = opt;
        this._batchCount = 0;
        this._messagesCount = 0;
        this._fileName = "";
        if (typeof opt.text !== "undefined" &&
            opt.parsing === true &&
            opt.text !== "") {
            this._lines = (0, utils_js_1.split)(opt.text).filter((line) => line.startsWith("MSH"));
        }
        else {
            this.set("FHS.7", (0, utils_js_1.createHL7Date)(new Date(), this._opt.date));
        }
    }
    /**
     * AAdd a Message or a Batch to the File
     * @remarks This adds a Message (MSH) output into the file batch.
     * If there is a Batch ("BHS") already part of this file, any new Message type will be added to the first found BHS regardless if the second Batch is added last.
     * @since 1.0.0
     * @param message The {@link Message} or {@link Batch} to add into the batch.
     */
    add(message) {
        this.setDirty();
        // if we are adding a message to a file
        if (message instanceof message_js_1.Message) {
            // and we already added a batch segment, we need to add it to the batch segment since we cannot add a batch and then a MSH segment.
            // That would violate HL7 specification.
            if (this._batchCount >= 1) {
                // get the first batch segment we find
                const batch = this._getFirstBatch();
                // update the count of the BTS segment by 1
                const seg = batch.getFirstSegment("BTS");
                seg.set(1, batch._messagesCount + 1);
                // add the message to the batch
                batch.add(message, batch._messagesCount + 1);
            }
            else {
                this._messagesCount = this._messagesCount + 1;
                this.children.push(message);
            }
        }
        else {
            // if there are already messages added before a batch
            if (this._messagesCount >= 1) {
                throw new exception_js_1.HL7ParserError("Unable to add a batch segment, since there is already messages added individually.");
            }
            this._batchCount = this._batchCount + 1;
            this.children.push(message);
        }
    }
    /**
     * Create a file to be stored.
     * @since 1.0.0
     * @param name Name of the file.
     */
    createFile(name) {
        const getFSHDate = this.get("FHS.7").toString();
        if (typeof name === "undefined") {
            throw new exception_js_1.HL7FatalError("Missing file name.");
        }
        if (constants_js_1.NAME_FORMAT.test(name)) {
            throw new exception_js_1.HL7FatalError("name must not contain certain characters: `!@#$%^&*()+\\-=\\[\\]{};':\"\\\\|,.<>\\/?~.");
        }
        if (typeof this._opt.location !== "undefined") {
            if (!node_fs_1.default.existsSync(this._opt.location)) {
                node_fs_1.default.mkdir(this._opt.location, { recursive: true }, () => { });
            }
            this._fileName = `hl7.${name}.${getFSHDate}.${this._opt.extension}`;
            node_fs_1.default.appendFile(node_path_1.default.join(this._opt.location, this._fileName), this.toString(), () => { });
        }
    }
    /**
     * End Batch
     * @remarks At the conclusion of building the file batch,
     * (Usually {@link add} method will be before this) will add the File Batch Trailing Segment (FTS) to the end.
     * If a message (MSH) is added after this,
     * that message (MSH) will get added to the first BHS found if there is one, otherwise it will just be added.
     * This might be typical inside a file output process.
     * @since 1.0.0
     */
    end() {
        const segment = this._addSegment("FTS");
        segment.set("1", this._batchCount + this._messagesCount);
    }
    /**
     * Get File name
     * @remarks Get File name going to be created.
     * @since 1.2.0
     */
    fileName() {
        return this._fileName;
    }
    /**
     * Get FHS Segment at Path
     * @since 1.0.0
     * @param path Could be 'FHS.7' or 7, and it shall get the same result.
     * @example
     * ```ts
     * const fileBatch = file.get('FHS.7')
     * ```
     * or
     * ```ts
     * const fileBatch = file.get(7)
     * ```
     */
    get(path) {
        return super.get(path);
    }
    /**
     * Get Messages within a submitted File Batch
     * @remarks This will parse the passed on "text"
     * in the contractor options and get all the messages (MSH) segments within it and return an array of them.
     * This will happen regardless of the depth of the segments.
     * @since 1.0.0
     * @example
     * ```ts
     * try {
     *  // parser the batch
     *  const parser = new FileBatch({ text: loadedMessage })
     *  // load the messages
     *  const allMessage = parser.messages()
     *  // loop messages
     *  allMessage.forEach((message: Message) => {
     *    const messageParsed = new Message({ text: message.toString() })
     *  })
     * } catch (e) {
     *   // error here
     * }
     * ```
     * @returns Returns an array of messages or a HL7ParserError will throw.
     */
    messages() {
        if (typeof this._lines !== "undefined" &&
            typeof this._opt.newLine !== "undefined") {
            const message = [];
            const re = new RegExp(`${this._opt.newLine}$`, "g");
            for (let i = 0; i < this._lines.length; i++) {
                message.push(new message_js_1.Message({ text: this._lines[i].replace(re, "") }));
            }
            return message;
        }
        throw new exception_js_1.HL7FatalError("No messages inside file segment.");
    }
    /**
     * Set Batch Segment at Path with a Value
     * @since 1.0.0
     * @example
     * ```ts
     * const batch = new Batch()
     * batch.set('BHS.7', '20231231')
     * ```
     * @param path Where you want to set in the segment
     * @param value The value.
     * It Can be an Array, String, or Boolean.
     * If the value is not set, you can chain this to expand the paths
     * @example
     * If the value is not used this can be employed:
     * ```ts
     * batch.set('BHS.3').set(0).set('BHS.3.1', 'abc');
     * ```
     */
    set(path, value) {
        return super.set(path, value);
    }
    /**
     * Start Batch
     * @since 1.0.0
     */
    start() {
        this.set("FHS.7", (0, utils_js_1.createHL7Date)(new Date()));
    }
    /** @internal */
    createChild(text, _index) {
        return new segment_js_1.Segment(this, text.trim());
    }
    /** @internal */
    read(path) {
        const segmentName = path.shift();
        if (path.length === 0) {
            const segments = this.children.filter((x) => x.name === segmentName);
            if (segments.length > 0) {
                return new segmentList_js_1.SegmentList(this, segments);
            }
        }
        else {
            if (typeof segmentName === "undefined") {
                throw new exception_js_1.HL7ParserError("Segment name is not defined.");
            }
            const segment = this._getFirstSegment(segmentName);
            if (typeof segment !== "undefined") {
                return segment.read(path);
            }
        }
        throw new exception_js_1.HL7FatalError("Unable to process the read function correctly.");
    }
    /** @internal */
    writeCore(path, value) {
        const segmentName = path.shift();
        if (typeof segmentName === "undefined") {
            throw new exception_js_1.HL7ParserError("Segment name is not defined.");
        }
        return this.writeAtIndex(path, value, 0, segmentName);
    }
    /** @internal **/
    _addSegment(path) {
        if (typeof path === "undefined") {
            throw new exception_js_1.HL7ParserError("Missing segment path.");
        }
        const preparedPath = this.preparePath(path);
        if (preparedPath.length !== 1) {
            throw new exception_js_1.HL7ParserError(`"Invalid segment ${path}."`);
        }
        return this.addChild(preparedPath[0]);
    }
    /** @internal */
    _getFirstBatch() {
        const children = this.children;
        for (let i = 0, l = children.length; i < l; i++) {
            if (children[i] instanceof batch_js_1.Batch) {
                return children[i];
            }
        }
        throw new exception_js_1.HL7FatalError("Unable to process _getFirstBatch.");
    }
    /** @internal */
    _getFirstSegment(name) {
        const children = this.children;
        for (let i = 0, l = children.length; i < l; i++) {
            const segment = children[i];
            if (segment.name === name) {
                return segment;
            }
        }
        throw new exception_js_1.HL7ParserError("Unable to process _getFirstSegment.");
    }
}
exports.FileBatch = FileBatch;
exports.default = FileBatch;
