// node-hl7-client/lib/cjs/utils/normalizedClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeClientOptions = normalizeClientOptions;
exports.normalizeClientListenerOptions = normalizeClientListenerOptions;
const exception_js_1 = require("./exception.js");
const utils_js_1 = require("./utils.js");
const DEFAULT_CLIENT_OPTS = {
    encoding: "utf-8",
    connectionTimeout: 0,
    maxAttempts: 10,
    maxConnectionAttempts: 10,
    maxTimeout: 10,
    retryHigh: 30000,
    retryLow: 1000,
};
const DEFAULT_LISTEN_CLIENT_OPTS = {
    autoConnect: true,
    maxAttempts: 10,
    maxConnectionAttempts: 10,
    waitAck: true,
};
/** @internal */
function normalizeClientOptions(raw) {
    const props = { ...DEFAULT_CLIENT_OPTS, ...raw };
    if (typeof props.host === "undefined" || props.host.length <= 0) {
        throw new exception_js_1.HL7FatalError("host is not defined or the length is less than 0.");
    }
    if (props.ipv4 === true && props.ipv6 === true) {
        throw new exception_js_1.HL7FatalError("ipv4 and ipv6 both can't be set to be both used exclusively.");
    }
    if (typeof props.host !== "string" &&
        props.ipv4 === false &&
        props.ipv6 === false) {
        throw new exception_js_1.HL7FatalError("host is not valid string.");
    }
    else if (typeof props.host === "string" &&
        props.ipv4 === true &&
        props.ipv6 === false) {
        if (!(0, utils_js_1.validIPv4)(props.host)) {
            throw new exception_js_1.HL7FatalError("host is not a valid IPv4 address.");
        }
    }
    else if (typeof props.host === "string" &&
        props.ipv4 === false &&
        props.ipv6 === true) {
        if (!(0, utils_js_1.validIPv6)(props.host)) {
            throw new exception_js_1.HL7FatalError("host is not a valid IPv6 address.");
        }
    }
    if (props.tls === true) {
        props.tls = {};
    }
    (0, utils_js_1.assertNumber)(props, "connectionTimeout", 0, 60000);
    (0, utils_js_1.assertNumber)(props, "maxTimeout", 1, 50);
    return props;
}
/** @internal */
function normalizeClientListenerOptions(client, raw) {
    const props = { ...DEFAULT_LISTEN_CLIENT_OPTS, ...raw };
    if (typeof props.port === "undefined") {
        throw new exception_js_1.HL7FatalError("port is not defined.");
    }
    if (typeof props.port !== "number") {
        throw new exception_js_1.HL7FatalError("port is not valid number.");
    }
    if (typeof props.retryHigh === "undefined") {
        props.retryHigh = client.retryHigh;
    }
    if (typeof props.retryLow === "undefined") {
        props.retryLow = client.retryLow;
    }
    (0, utils_js_1.assertNumber)(props, "maxAttempts", 1, 50);
    (0, utils_js_1.assertNumber)(props, "maxConnectionAttempts", 1, 50);
    (0, utils_js_1.assertNumber)(props, "port", 1, 65353);
    return props;
}
