// node-hl7-client/lib/cjs/utils/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDeferred = exports.split = exports.randomString = exports.validIPv6 = exports.validIPv4 = exports.padHL7Date = exports.isHL7String = exports.isHL7Number = exports.isFile = exports.isBatch = exports.expBackoff = exports.escapeForRegExp = exports.decodeHexString = exports.createHL7Date = exports.assertNumber = void 0;
/**
 * Assert Number on a Property
 * @since 1.0.0
 * @param props Property Object
 * @param name Property Name
 * @param min Min Number
 * @param max Max Number
 */
const assertNumber = (props, name, min, max) => {
    const val = props[name];
    if (isNaN(val) ||
        !Number.isFinite(val) ||
        val < min ||
        (max != null && val > max)) {
        throw new TypeError(max != null
            ? `${name} must be a number (${min}, ${max}).`
            : `${name} must be a number >= ${min}.`);
    }
};
exports.assertNumber = assertNumber;
/**
 * Create a valid HL7 Date.
 * @remarks Custom for this package and based of HL7 specification.
 * @since 1.0.0
 * @param date
 * @param length
 */
const createHL7Date = (date, length) => {
    switch (length) {
        case "8":
            return `${date.getFullYear()}${(0, exports.padHL7Date)(date.getMonth() + 1, 2)}${(0, exports.padHL7Date)(date.getDate(), 2)}`;
        case "12":
            return `${date.getFullYear()}${(0, exports.padHL7Date)(date.getMonth() + 1, 2)}${(0, exports.padHL7Date)(date.getDate(), 2)}${(0, exports.padHL7Date)(date.getHours(), 2)}${(0, exports.padHL7Date)(date.getMinutes(), 2)}`;
        case "14":
        default:
            return `${date.getFullYear()}${(0, exports.padHL7Date)(date.getMonth() + 1, 2)}${(0, exports.padHL7Date)(date.getDate(), 2)}${(0, exports.padHL7Date)(date.getHours(), 2)}${(0, exports.padHL7Date)(date.getMinutes(), 2)}${(0, exports.padHL7Date)(date.getSeconds(), 2)}`;
    }
};
exports.createHL7Date = createHL7Date;
/**
 * Decode Hex String
 * @since 1.0.0
 * @param value
 */
const decodeHexString = (value) => {
    const result = new Array(value.length / 2);
    for (let i = 0; i < value.length; i += 2) {
        result[i / 2] = String.fromCharCode(parseInt(value.slice(i, i + 2), 16));
    }
    return result.join("");
};
exports.decodeHexString = decodeHexString;
/**
 * Escape for RegEx Expressing
 * @since 1.0.0
 * @param value
 */
const escapeForRegExp = (value) => {
    return value.replace(/[-/\\^$*+?.()|[\]{}]/g, "\\$&");
};
exports.escapeForRegExp = escapeForRegExp;
/**
 * Calculate exponential backoff/retry delay.
 * Where attempts >= 1, exp > 1
 * @example expBackoff(1000, 30000, attempts)
 *   ---------------------------------
 *    attempts | possible delay
 *   ----------+----------------------
 *        1    | 1000 to 2000
 *        2    | 1000 to 4000
 *        3    | 1000 to 8000
 *        4    | 1000 to 16000
 *        5    | 1000 to 30000
 *   ---------------------------------
 * Attempts required before max delay is possible = Math.ceil(Math.log(high/step) / Math.log(exp))
 * @since 1.0.0
 * @param step
 * @param high
 * @param attempts
 * @param exp
 */
const expBackoff = (step, high, attempts, exp = 2) => {
    const slots = Math.ceil(Math.min(high / step, Math.pow(exp, attempts)));
    const max = Math.min(slots * step, high);
    return Math.floor(Math.random() * (max - step) + step);
};
exports.expBackoff = expBackoff;
/**
 * Check to see if the message starts with Batch (BHS) header segment.
 * @since 1.0.0
 * @param message
 */
const isBatch = (message) => {
    const lines = (0, exports.split)(message).filter((line) => line.startsWith("MSH")).length > 1 || false;
    if (lines) {
        return true;
    }
    else if (message.startsWith("MSH") && !lines) {
        return false;
    }
    return message.startsWith("BHS");
};
exports.isBatch = isBatch;
/**
 * Check to see if the message starts with a File Batch (FHS) header segment.
 * @param message
 */
const isFile = (message) => {
    return message.startsWith("FHS");
};
exports.isFile = isFile;
/**
 * Is Number
 * @remarks Custom for this package.
 * @since 1.0.0
 * @param value
 */
const isHL7Number = (value) => {
    value = typeof value === "string" ? parseInt(value) : value;
    return !isNaN(value) || !Number.isFinite(value);
};
exports.isHL7Number = isHL7Number;
/**
 * Is String
 * @remarks Custom for this package.
 * @since 1.0.0
 * @param value
 */
const isHL7String = (value) => typeof value === "string";
exports.isHL7String = isHL7String;
/**
 * HL7 Padding for Date
 * @since 1.0.0
 * @param n
 * @param width
 * @param z
 */
const padHL7Date = (n, width, z = "0") => {
    const s = n.toString();
    return s.length >= width ? s : new Array(width - s.length + 1).join(z) + s;
};
exports.padHL7Date = padHL7Date;
/**
 * Valid IPv4 Checker
 * @since 1.0.0
 * @param ip
 */
const validIPv4 = (ip) => {
    const ipv4Regex = /^(\d{1,3}\.){3}\d{1,3}$/;
    if (ipv4Regex.test(ip)) {
        return ip.split(".").every((part) => parseInt(part) <= 255);
    }
    return false;
};
exports.validIPv4 = validIPv4;
/**
 * Valid IPv6 Checker
 * @since 1.0.0
 * @param ip
 */
const validIPv6 = (ip) => {
    const ipv6Regex = /^([\da-f]{1,4}:){7}[\da-f]{1,4}$/i;
    if (ipv6Regex.test(ip)) {
        return ip.split(":").every((part) => part.length <= 4);
    }
    return false;
};
exports.validIPv6 = validIPv6;
/**
 * Generate a random string
 * @since 1.0.0
 * @param length
 */
const randomString = (length = 20) => {
    let result = "";
    const characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_";
    const charactersLength = characters.length;
    let counter = 0;
    while (counter < length) {
        result += characters.charAt(Math.floor(Math.random() * charactersLength));
        counter += 1;
    }
    return result;
};
exports.randomString = randomString;
/**
 * Split the message.
 * @remarks Split the message into its parts.
 * @since 1.0.0
 * @param data
 * @param segments
 */
const split = (data, segments = []) => {
    const getSegIndex = [
        ...getSegIndexes(["FHS", "BHS", "MSH", "BTS", "FTS"], data),
    ];
    getSegIndex.sort((a, b) => parseInt(a) - parseInt(b));
    for (let i = 0; i < getSegIndex.length; i++) {
        const start = parseInt(getSegIndex[i]);
        let end = parseInt(getSegIndex[i + 1]);
        if (i + 1 === getSegIndex.length) {
            end = data.length;
        }
        segments.push(data.slice(start, end));
    }
    return segments;
};
exports.split = split;
/**
 * Get Segment Indexes
 * @remarks Helper for {@link split}
 * @since 1.0.0
 * @param names
 * @param data
 * @param list
 */
const getSegIndexes = (names, data, list = []) => {
    for (let i = 0; i < names.length; i++) {
        const regex = new RegExp(`(\\n|\\r|)(${names[i]})\\|`, "g");
        let m;
        while ((m = regex.exec(data)) != null) {
            const s = m[0];
            if (s.includes("\r\n")) {
                m.index = m.index + 2;
            }
            else if (s.includes("\n")) {
                m.index++;
            }
            else if (s.includes("\r")) {
                m.index++;
            }
            if (m.index !== null) {
                list.push(m.index.toString());
            }
        }
    }
    return list;
};
/**
 * Create Deferred
 * @since 2.0.0
 * @param noUncaught
 */
const createDeferred = (noUncaught) => {
    const dfd = {};
    dfd.promise = new Promise((resolve, reject) => {
        dfd.resolve = resolve;
        dfd.reject = reject;
    });
    /* istanbul ignore next */
    if (noUncaught === false) {
        dfd.promise.catch(() => { });
    }
    return dfd;
};
exports.createDeferred = createDeferred;
