// node-hl7-client/lib/cjs/hl7.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HL7_SPEC_BASE = exports.HL7_2_8 = exports.HL7_2_7_1 = exports.HL7_2_7 = exports.HL7_2_6 = exports.HL7_2_5_1 = exports.HL7_2_5 = exports.HL7_2_4 = exports.HL7_2_3_1 = exports.HL7_2_3 = exports.HL7_2_2 = exports.HL7_2_1 = void 0;
/** HL7 Class **/
var _2_1_js_1 = require("./specification/2.1.js");
Object.defineProperty(exports, "HL7_2_1", { enumerable: true, get: function () { return _2_1_js_1.HL7_2_1; } });
var _2_2_js_1 = require("./specification/2.2.js");
Object.defineProperty(exports, "HL7_2_2", { enumerable: true, get: function () { return _2_2_js_1.HL7_2_2; } });
var _2_3_js_1 = require("./specification/2.3.js");
Object.defineProperty(exports, "HL7_2_3", { enumerable: true, get: function () { return _2_3_js_1.HL7_2_3; } });
var _2_3_1_js_1 = require("./specification/2.3.1.js");
Object.defineProperty(exports, "HL7_2_3_1", { enumerable: true, get: function () { return _2_3_1_js_1.HL7_2_3_1; } });
var _2_4_js_1 = require("./specification/2.4.js");
Object.defineProperty(exports, "HL7_2_4", { enumerable: true, get: function () { return _2_4_js_1.HL7_2_4; } });
var _2_5_js_1 = require("./specification/2.5.js");
Object.defineProperty(exports, "HL7_2_5", { enumerable: true, get: function () { return _2_5_js_1.HL7_2_5; } });
var _2_5_1_js_1 = require("./specification/2.5.1.js");
Object.defineProperty(exports, "HL7_2_5_1", { enumerable: true, get: function () { return _2_5_1_js_1.HL7_2_5_1; } });
var _2_6_js_1 = require("./specification/2.6.js");
Object.defineProperty(exports, "HL7_2_6", { enumerable: true, get: function () { return _2_6_js_1.HL7_2_6; } });
var _2_7_js_1 = require("./specification/2.7.js");
Object.defineProperty(exports, "HL7_2_7", { enumerable: true, get: function () { return _2_7_js_1.HL7_2_7; } });
var _2_7_1_js_1 = require("./specification/2.7.1.js");
Object.defineProperty(exports, "HL7_2_7_1", { enumerable: true, get: function () { return _2_7_1_js_1.HL7_2_7_1; } });
var _2_8_js_1 = require("./specification/2.8.js");
Object.defineProperty(exports, "HL7_2_8", { enumerable: true, get: function () { return _2_8_js_1.HL7_2_8; } });
var specification_js_1 = require("./specification/specification.js");
Object.defineProperty(exports, "HL7_SPEC_BASE", { enumerable: true, get: function () { return specification_js_1.HL7_SPEC_BASE; } });
